nft_ingester/src/program_transformers/bubblegum/db.rs

```diff
@@ -441,3 +441,327 @@ where
+        royalty_target: Set(royalty_target),
+        royalty_amount: Set(royalty_amount),
+        asset_data: Set(Some(id)),
+        slot_updated: Set(Some(slot_updated)),
+        ..Default::default()
+    };
+
+    // Upsert asset table base info.
+    let query = asset::Entity::insert(asset_model)
+        .on_conflict(
+            OnConflict::columns([asset::Column::Id])
+                .update_columns([
+                    asset::Column::OwnerType,
+                    asset::Column::Frozen,
+                    asset::Column::SpecificationVersion,
+                    asset::Column::SpecificationAssetClass,
+                    asset::Column::RoyaltyTargetType,
+                    asset::Column::RoyaltyTarget,
+                    asset::Column::RoyaltyAmount,
+                    asset::Column::AssetData,
+                    asset::Column::SlotUpdated,
+                ])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+
+    txn.execute(query)
+        .await
+        .map_err(|db_err| IngesterError::AssetIndexError(db_err.to_string()))?;
+
+    Ok(())
+}
+
+pub async fn asset_should_be_updated<T>(
+    txn: &T,
+    id: Vec<u8>,
+    seq: Option<i64>,
+) -> Result<bool, IngesterError>
+where
+    T: ConnectionTrait + TransactionTrait,
+{
+    if let Some(asset) = asset::Entity::find_by_id(id).one(txn).await? {
+        if let Some(update_metadata_seq) = asset.update_metadata_seq {
+            if update_metadata_seq == 0 {
+                // Asset was decompressed and should no longer be updated by Bubblegum program
+                // transformers.
+                return Ok(false);
+            } else if let Some(seq) = seq {
+                if seq < update_metadata_seq {
+                    // Asset was updated by an `update_metadata` so should not be modified by this
+                    // sequence number.
+                    return Ok(false);
+                }
+            }
+        }
+    };
+    Ok(true)
+}
+
+pub async fn upsert_asset_with_update_metadata_seq<T>(
+    txn: &T,
+    id: Vec<u8>,
+    seq: i64,
+) -> Result<(), IngesterError>
+where
+    T: ConnectionTrait + TransactionTrait,
+{
+    let model = asset::ActiveModel {
+        id: Set(id),
+        update_metadata_seq: Set(Some(seq)),
+        ..Default::default()
+    };
+
+    let mut query = asset::Entity::insert(model)
+        .on_conflict(
+            OnConflict::column(asset::Column::Id)
+                .update_columns([asset::Column::UpdateMetadataSeq])
+                .to_owned(),
+        )
+        .build(DbBackend::Postgres);
+
+    query.sql = format!(
+        "{} WHERE excluded.update_metadata_seq >= asset.update_metadata_seq OR asset.update_metadata_seq IS NULL",
+        query.sql
+    );
+
+    txn.execute(query)
+        .await
+        .map_err(|db_err| IngesterError::StorageWriteError(db_err.to_string()))?;
+
+    Ok(())
+}
+
+pub async fn unprotected_upsert_creators<T>(
```

Comment:
this method seems to be doing updates, so shouldn't it be protected?

Reply:
This is part of the upserts that were not inherently protected by their own SQL logic, but blanket protected by `update_metadata_seq` that we now know has a concurrency issue.

Reply:
Fixed in https://github.com/metaplex-foundation/digital-asset-rpc-infrastructure/pull/134